src/libfuncs/nullable.rs

```diff
@@ -173,6 +173,26 @@ where
     Ok(())
 }
 
+/// Generate MLIR operations for the `match_nullable_snapshot` libfunc.
+#[allow(clippy::too_many_arguments)]
+fn build_match_nullable_snapshot<'ctx, 'this, TType, TLibfunc>(
+    context: &'ctx Context,
+    registry: &ProgramRegistry<TType, TLibfunc>,
+    entry: &'this Block<'ctx>,
+    location: Location<'ctx>,
+    helper: &LibfuncHelper<'ctx, 'this>,
+    metadata: &mut MetadataStorage,
+    info: &SignatureAndTypeConcreteLibfunc,
+) -> Result<()>
+where
+    TType: GenericType,
+    TLibfunc: GenericLibfunc,
+    <TType as GenericType>::Concrete: TypeBuilder<TType, TLibfunc, Error = CoreTypeBuilderError>,
+    <TLibfunc as GenericLibfunc>::Concrete: LibfuncBuilder<TType, TLibfunc, Error = Error>,
+{
+    build_match_nullable(context, registry, entry, location, helper, metadata, info)
```

Comment:
I'm not 100% sure about this, but since the input of the libfunc is a `@Nullable<T>`, does it make since to say it's equivalent to a `Nullable<T>` since `snapshot_take` just copies the pointer to the value in the MLIR implementation? Same can be said about `Box<T>` and `Box<@T>`.

Reply:
Yes, `snapshot_take` is (or should be) equivalent to a `.clone()` in rust, not a plain copy since otherwise we'd get double free and other undesireable side effects. Because of this, any type containing snapshots should be internally equivalent to the same type but with snapshots removed.

TLDR: It should be correct.